scene/gui/text_edit.cpp

```diff
@@ -4179,6 +4384,198 @@ void TextEdit::get_breakpoints(List<int> *p_breakpoints) const {
 	}
 }
 
+void TextEdit::set_line_as_hidden(int p_line, bool p_hidden) {
+
+	ERR_FAIL_INDEX(p_line, text.size());
+	if (is_hiding_enabled() || !p_hidden)
+		text.set_hidden(p_line, p_hidden);
+	update();
+}
+
+bool TextEdit::is_line_hidden(int p_line) const {
+
+	ERR_FAIL_INDEX_V(p_line, text.size(), false);
+	return text.is_hidden(p_line);
+}
+
+void TextEdit::fold_all_lines() {
+
+	for (int i = 0; i < text.size(); i++) {
+		fold_line(i);
+	}
+	_update_scrollbars();
+	update();
+}
+
+void TextEdit::unhide_all_lines() {
+
+	for (int i = 0; i < text.size(); i++) {
+		text.set_hidden(i, false);
+	}
+	_update_scrollbars();
+	update();
+}
+
+int TextEdit::num_lines_from(int p_line_from, int unhidden_amount) const {
+
+	// returns the number of hidden and unhidden lines from p_line_from to p_line_from + amount of visible lines
+	ERR_FAIL_INDEX_V(p_line_from, text.size(), ABS(unhidden_amount));
+
+	if (!is_hiding_enabled())
+		return unhidden_amount;
+	int num_visible = 0;
+	int num_total = 0;
+	if (unhidden_amount >= 0) {
+		for (int i = p_line_from; i < text.size(); i++) {
+			num_total++;
+			if (!is_line_hidden(i))
+				num_visible++;
+			if (num_visible >= unhidden_amount)
+				break;
+		}
+	} else {
+		unhidden_amount = ABS(unhidden_amount);
+		for (int i = p_line_from; i >= 0; i--) {
+			num_total++;
+			if (!is_line_hidden(i))
+				num_visible++;
+			if (num_visible >= unhidden_amount)
+				break;
+		}
+	}
+	return num_total;
+}
+
+int TextEdit::get_whitespace_level(int p_line) const {
+
+	ERR_FAIL_INDEX_V(p_line, text.size(), 0);
+
+	// counts number of tabs and spaces before line starts
+	int whitespace_count = 0;
+	int line_length = text[p_line].size();
+	for (int i = 0; i < line_length - 1; i++) {
+		if (text[p_line][i] == '\t') {
+			whitespace_count++;
+		} else if (text[p_line][i] == ' ') {
+			whitespace_count++;
+		} else if (text[p_line][i] == '#') {
+			break;
+		} else {
+			break;
+		}
+	}
+	return whitespace_count;
+}
+
+bool TextEdit::can_fold(int p_line) const {
+
+	ERR_FAIL_INDEX_V(p_line, text.size(), false);
+	if (!is_hiding_enabled())
+		return false;
+	if (p_line + 1 >= text.size())
+		return false;
+	if (text[p_line].size() == 0)
+		return false;
+	if (is_folded(p_line))
+		return false;
+	if (is_line_hidden(p_line))
+		return false;
+
+	int start_indent = get_whitespace_level(p_line);
+
+	for (int i = p_line + 1; i < text.size(); i++) {
+		if (text[i].size() == 0)
+			continue;
+		int next_indent = get_whitespace_level(i);
+		if (next_indent > start_indent)
+			return true;
+		else
+			return false;
+	}
+
+	return false;
+}
+
+bool TextEdit::is_folded(int p_line) const {
+
+	ERR_FAIL_INDEX_V(p_line, text.size(), false);
+	if (p_line + 1 >= text.size() - 1)
```

Comment:
This line should be `p_line + 1 >= text.size()` else you can't unfold the last line.

Reply:
The last line isn't able to be folded. I determine whether a line is folded by seeing if this line isn't hidden and the next one is hidden, so the next line needs to be less than or equal to `text.size() - 1` for that check.